File: TelerikUIExamples/TelerikUIExamples/ExampleViewController.m
//
//  ExampleViewController.m
//  TelerikUIExamples
//
//  Copyright (c) 2013 Telerik. All rights reserved.
//

#import <objc/message.h>
#import "ExampleViewController.h"
#import "OptionInfo.h"
#import "SettingsViewController.h"
#import "OptionSection.h"

@implementation ExampleViewController
{
    NSMutableArray *_options;
    NSMutableArray *_sections;
    CGFloat _headerHeight;
    CGFloat _offset;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        _options = [[NSMutableArray alloc] init];
        _selectedOptions = [[NSMutableSet alloc] init];
        _headerHeight = 0;
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor whiteColor];
    self.navigationController.navigationBar.translucent = NO;
    
    [self updateHeaderHeight];
    
    UIUserInterfaceIdiom idiom = [[UIDevice currentDevice] userInterfaceIdiom];
    if (_options.count > 0 || _sections) {
        if (idiom == UIUserInterfaceIdiomPad) {
            [self loadIPadLayout];
        }
        else {
            [self loadIPhoneLayout];
        }
    }
    
    [self updateLayoutConstraints];
}

- (void)willMoveToParentViewController:(UIViewController *)parent
{
    [super willMoveToParentViewController:parent];
    if (parent == nil) {
        self.navigationController.navigationBar.translucent = YES;
    }
}

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    [self updateLayoutConstraints];
}

- (void)updateHeaderHeight
{
    UINavigationBar *navigationBar = self.navigationController.navigationBar;
    if (navigationBar.translucent) {
        UIApplication *app = [UIApplication sharedApplication];
        BOOL isLandscape = UIInterfaceOrientationIsLandscape(app.statusBarOrientation);
        UIUserInterfaceIdiom idiom = [[UIDevice currentDevice] userInterfaceIdiom];
        CGFloat version = [[UIDevice currentDevice].systemVersion floatValue];
        if (idiom == UIUserInterfaceIdiomPad && version >= 8.0) {
            _headerHeight = navigationBar.intrinsicContentSize.height + app.statusBarFrame.size.height;
        }
        else {
            _headerHeight = navigationBar.intrinsicContentSize.height + (isLandscape ? app.statusBarFrame.size.width : app.statusBarFrame.size.height);
        }
    }
}

- (void)updateLayoutConstraints
{
    _exampleBounds = self.view.bounds;
    
    [self updateHeaderHeight];
    
    _exampleBounds.origin.y += _headerHeight;
    _exampleBounds.size.height -= _headerHeight;

    if (_offset > 0) {
        _exampleBounds.origin.y += _offset;
        _exampleBounds.size.height -= _offset;
    }
    
    UIUserInterfaceIdiom idiom = [[UIDevice currentDevice] userInterfaceIdiom];
    if (idiom == UIUserInterfaceIdiomPad) {
        _exampleBoundsWithInset = CGRectInset(_exampleBounds, 30, 30);
    }
    else {
        _exampleBoundsWithInset = CGRectInset(_exampleBounds, 10, 10);
    }
}

- (void)loadIPhoneLayout
{
    if (_sections == nil && _options.count == 1) {
        OptionInfo *info = _options[0];
        _settingsButton = [[UIBarButtonItem alloc] initWithTitle:info.optionText
                                                           style:UIBarButtonItemStylePlain
                                                          target:self
                                                          action:@selector(optionTouched)];
    }
    else {
        _settingsButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"menu"]
                                                           style:UIBarButtonItemStylePlain
                                                          target:self
                                                          action:@selector(settingsTouched)];
    }
    self.navigationItem.rightBarButtonItem = _settingsButton;
}

- (void)loadIPadLayout
{
    CGSize desiredSize = CGSizeZero;
    _offset = 0;
    if (_sections == nil && _options.count == 1) {
        OptionInfo *info = _options[0];
        UIButton *button = [UIButton buttonWithType:UIButtonTypeRoundedRect];
        [button addTarget:self action:@selector(optionTouched) forControlEvents:UIControlEventTouchDown];
        [button setTitle:info.optionText forState:UIControlStateNormal];
        desiredSize = [button sizeThatFits:self.view.bounds.size];
        button.frame = CGRectMake(30, 10 + _headerHeight, desiredSize.width, desiredSize.height);
        [self.view addSubview:button];
        _offset = 10 + desiredSize.height + 10;
    }
    else if (_options.count >= 3 || _sections) {
        _settingsButton = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"menu"] style:UIBarButtonItemStylePlain target:self action:@selector(settingsTouched)];
        self.navigationItem.rightBarButtonItem = _settingsButton;
    }
    else {
        UISegmentedControl *segmented = [[UISegmentedControl alloc] init];
        int i = 0;
        for (OptionInfo *option in _options) {
            [segmented insertSegmentWithTitle:option.optionText atIndex:i++ animated:NO];
        }
        desiredSize = [segmented sizeThatFits:self.view.bounds.size];
        segmented.frame = CGRectMake(10, 10 + _headerHeight, desiredSize.width, desiredSize.height);
        segmented.selectedSegmentIndex = self.selectedOption;
        [segmented addTarget:self action:@selector(optionSelected:) forControlEvents:UIControlEventValueChanged];
        [self.view addSubview:segmented];
        _offset = 10 + desiredSize.height + 10;
    }
}

- (CGRect)exampleBounds
{
    return _exampleBounds;
}

- (void)addOption:(NSString*)text selector:(SEL)selector
{
    [_options addObject:[[OptionInfo alloc] initWithText:text selector:selector]];
}

- (void) addOption:(OptionInfo *)option
{
    [_options addObject:option];
}

- (void)addOption:(NSString*)text selector:(SEL)selector inSection:(NSString*)sectionName
{
    [self addOption:[[OptionInfo alloc] initWithText:text selector:selector] inSection:sectionName];
}

- (void)addOption:(OptionInfo *)option inSection:(NSString*)sectionName
{
    if (!_sections) {
        _sections = [NSMutableArray new];
    }
    for (OptionSection *section in _sections) {
        if ([section.title isEqualToString:sectionName]) {
            [section.items addObject:option];
            return;
        }
    }
    OptionSection *section = [[OptionSection alloc] initWithTitle:sectionName];
    [section.items addObject:option];
    [_sections addObject:section];
}

- (NSArray *)options
{
    return _options;
}

- (NSArray *)sections
{
    return _sections;
}

- (void)setSelectedOption:(NSInteger)selectedOption inSection:(NSInteger)section
{
    OptionSection *sec = _sections[section];
    sec.selectedOption = selectedOption;
}

- (void)optionSelected:(UISegmentedControl*)sender
{
    if (sender.selectedSegmentIndex >= 0) {
        self.selectedOption = sender.selectedSegmentIndex;
        OptionInfo *info = _options[sender.selectedSegmentIndex];
        objc_msgSend(self, info.selector);
    }
}

- (void)optionTouched
{
    OptionInfo *info = _options[0];
    objc_msgSend(self, info.selector);
}

- (void)settingsTouched
{
    SettingsViewController *settings = [[SettingsViewController alloc] initWithOptions:_options];
    settings.sections = _sections;
    settings.example = self;
    settings.selectedOption = self.selectedOption;
    
    UIUserInterfaceIdiom idiom = [[UIDevice currentDevice] userInterfaceIdiom];
    if (idiom == UIUserInterfaceIdiomPad) {
        if (_popover && [_popover isPopoverVisible]) {
            [_popover dismissPopoverAnimated:YES];
            
            return;
        }
        
        _popover = [[UIPopoverController alloc] initWithContentViewController:settings];
        CGRect settingsRect = settings.view.bounds;
        [settings.table sizeToFit];
        CGSize size = CGSizeMake(MIN(settingsRect.size.width, settingsRect.size.height) / 2., settings.table.contentSize.height);
        _popover.popoverContentSize = size;
        [_popover presentPopoverFromBarButtonItem:_settingsButton permittedArrowDirections:UIPopoverArrowDirectionUp animated:YES];
    }
    else {
        [self.navigationController pushViewController:settings animated:YES];
    }
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    if (_popover && [_popover isPopoverVisible]) {
        [_popover dismissPopoverAnimated:NO];
    }
}

@end

File: TelerikUIExamplesInSwift2.0/TelerikUIExamplesInSwift/ExampleViewController.swift
//
//  ExampleViewController.swift
//  Swift Examples
//
//  Copyright (c) 2014 Telerik. All rights reserved.
//

import Foundation

class ExampleViewController: UIViewController {
    
    let options : NSMutableArray = NSMutableArray()
    let selectedOptions : NSMutableSet = NSMutableSet()
    var selectedOption : NSInteger = 0
    var popover: UIPopoverController?
    var settingsButton : UIBarButtonItem?
    var exampleBounds = CGRect.zeroRect
    var exampleBoundsWithInset = CGRect.zeroRect
    var sections = NSMutableArray()
    var headerHeight:CGFloat = 0
    var offset:CGFloat = 0
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.view.backgroundColor = UIColor.whiteColor()
        
        self.updateHeaderHeight()
        
        let idiom = UIDevice.currentDevice().userInterfaceIdiom as UIUserInterfaceIdiom
        if sections.count > 0 || options.count > 0 {
            if (idiom == UIUserInterfaceIdiom.Pad) {
                self.loadIPadLayout()
            }
            else {
                self.loadIPhoneLayout()
            }
        }
        
        self.updateLayoutConstraints()
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        self.updateLayoutConstraints()
    }
    
    func updateHeaderHeight() {
        if  self.navigationController != nil {
            let navigationBar = self.navigationController!.navigationBar
            if navigationBar.translucent {
                let app = UIApplication.sharedApplication()
                let isLandscape = UIInterfaceOrientationIsLandscape(app.statusBarOrientation)
                self.headerHeight = navigationBar.intrinsicContentSize().height + (isLandscape ? app.statusBarFrame.size.width : app.statusBarFrame.size.height)
            }
        }
    }
    
    func updateLayoutConstraints() {
        
        self.exampleBounds = self.view.bounds

        self.updateHeaderHeight()

        self.exampleBounds.origin.y += headerHeight
        self.exampleBounds.size.height -= headerHeight
    
        if (self.offset > 0) {
            self.exampleBounds.origin.y += self.offset
            self.exampleBounds.size.height -= self.offset
        }
    
        let idiom = UIDevice.currentDevice().userInterfaceIdiom
        if idiom == UIUserInterfaceIdiom.Pad {
            self.exampleBoundsWithInset = CGRectInset(self.exampleBounds, 30, 30);
        }
        else {
            self.exampleBoundsWithInset = CGRectInset(self.exampleBounds, 10, 10);
        }
    }
    
    func loadIPhoneLayout() {
        if sections.count == 0 && options.count == 1 {
           let info = options[0] as! OptionInfo
           self.settingsButton = UIBarButtonItem(title: info.optionText, style: UIBarButtonItemStyle.Plain, target: self, action: "optionTouched")
        }
        else {
            let image = UIImage(named: "menu")
            self.settingsButton = UIBarButtonItem(image: image, style: UIBarButtonItemStyle.Plain, target: self, action: "settingsTouched")
        }
        self.navigationItem.rightBarButtonItem = self.settingsButton
    }
    
    func loadIPadLayout() {
        self.offset = 0
        var desiredSize: CGSize = CGSizeZero
        if sections.count == 0 && options.count == 1 {
            let info  = options[0] as! OptionInfo
            let button = UIButton(type: UIButtonType.System) as UIButton
            button.addTarget(self, action: "optionTouched", forControlEvents: UIControlEvents.TouchDown)
            button.setTitle(info.optionText, forState: UIControlState.Normal)
            desiredSize = button.sizeThatFits(self.view.bounds.size)
            button.frame = CGRectMake(30, 10 + self.headerHeight, desiredSize.width, desiredSize.height)
            self.view.addSubview(button)
            self.offset = 10 + desiredSize.height + 10
        }
        else if options.count >= 3 || sections.count > 0 {
            settingsButton = UIBarButtonItem(image: UIImage(named: "menu"), style: UIBarButtonItemStyle.Plain, target: self, action: "settingsTouched")
            self.navigationItem.rightBarButtonItem = settingsButton
        }
        else {
            let segmented = UISegmentedControl()
            var i: Int = 0
            for o in options {
                let option = o as! OptionInfo
                segmented.insertSegmentWithTitle(option.optionText, atIndex: i++, animated: false)
            }
            desiredSize = segmented.sizeThatFits(self.view.bounds.size)
            segmented.frame = CGRectMake(10, 10 + headerHeight, desiredSize.width, desiredSize.height)
            segmented.selectedSegmentIndex = self.selectedOption
            segmented.addTarget(self, action: "optionSelected:", forControlEvents: UIControlEvents.ValueChanged)
            self.view.addSubview(segmented)
            self.offset = 10 + desiredSize.height + 10
        }
    }
    
    func addOption(text:String, selector: Optional<() -> ()>) {
        options.addObject(OptionInfo(Text: text, Selector: selector))
    }
    
    func addOption(option: OptionInfo) {
        options.addObject(option)
    }
    
    func addOption(text:String, inSection:String, selector: Optional<() -> ()>) {
        addOption(OptionInfo(Text:text, Selector:selector), inSection:inSection)
    }
    
    func addOption(option: OptionInfo, inSection:String)
    {
        for section in sections {
            let sec = section as! OptionSection
            if sec.title == inSection {
                sec.items.addObject(option)
                return;
            }
        }
        let section = OptionSection(Text: inSection)
        section.items.addObject(option)
        sections.addObject(section)
    }
    
    func optionSelected(sender: UISegmentedControl) {
        if (sender.selectedSegmentIndex >= 0) {
            self.selectedOption = sender.selectedSegmentIndex
            let info = options[sender.selectedSegmentIndex] as! OptionInfo
            info.selector?()
        }
    }
    
    func setSelectedOption(index: Int, section: Int) {
        let sec = sections[section] as! OptionSection
        sec.selectedOption = index
    }
    
    func optionTouched() {
        let info = options[0] as! OptionInfo
        info.selector?()
    }
  
    func settingsTouched() {
        let settings = SettingsViewController()
        settings.options = options
        settings.example = self
        settings.sections = sections
        settings.selectedOption = self.selectedOption
        
        let idiom = UIDevice.currentDevice().userInterfaceIdiom as UIUserInterfaceIdiom
        if (idiom == UIUserInterfaceIdiom.Pad) {
            if (popover != nil && popover!.popoverVisible) {
                popover!.dismissPopoverAnimated(true)
                return
            }
            
            popover = UIPopoverController(contentViewController: settings)
            let rect = settings.view.bounds
            settings.table.sizeToFit()
            let size = CGSizeMake(min(rect.size.width, rect.size.height) / 2, settings.table.contentSize.height) as CGSize
            popover!.popoverContentSize = size
            popover!.presentPopoverFromBarButtonItem(settingsButton!, permittedArrowDirections: UIPopoverArrowDirection.Up, animated: true)
        }
        else {
            self.navigationController?.pushViewController(settings, animated: true)
        }
    }
    
    override func viewWillDisappear(animated: Bool) {
        if (popover != nil && popover!.popoverVisible) {
            popover!.dismissPopoverAnimated(true)
        }
    }
}
